src/main/java/org/opentripplanner/transit/model/network/SubMode.java

```diff
@@ -0,0 +1,122 @@
+package org.opentripplanner.transit.model.network;
+
+import java.io.Serializable;
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class wrap a string and create an index for it, so we can use BitSet for matching a
+ * trip netexSubMode. SubModes are used in trip filtering for every request.
+ * <p>
+ * Naming, this class is named SubMode, not NetexSubMode because we want to migrate gtfsRouteType
+ * into the same concept.
+ * <p>
+ * This class is thread-safe, and the performance overhead should affect the graph built time,
+ * but not routing.
+ */
+public record SubMode(String name, int index) implements Serializable {
+  private static final int NONE_EXISTING_SUB_MODE_INDEX = 1_000_000;
+
+  private static final Logger LOG = LoggerFactory.getLogger(SubMode.class);
+
+  /**
+   * Note! This cache all sub-modes used in the static scheduled transit data. When
+   * serialized the static fields need to be explicit serialized, it is not enough
+   * to serialize the instances.
+   */
+  private static final Map<String, SubMode> ALL = new ConcurrentHashMap<>();
+  private static final AtomicInteger COUNTER = new AtomicInteger(0);
+
+  public static final SubMode UNKNOWN = getOrBuildAndCacheForever("unknown");
+
+  /**
+   * This method is safe to use in a request scope. Usually you want to fetch an instant to
+   * pass in as a request parameter. This method will fetch an existing object - not creating
+   * a duplicate object, if at least one Route, Trip or Stop has this submode. If it does not
+   * exist in the model a new is created, but not put on the "deduplication" map.
+   */
+  public static SubMode of(String name) {
+    if (name == null) {
+      return UNKNOWN;
+    }
+    var subMode = ALL.get(name);
+    return subMode != null ? subMode : new SubMode(name, NONE_EXISTING_SUB_MODE_INDEX);
```

Comment:
Shouldn't it be NON_EXISTING_SUB_MODE_INDEX?

Reply:
Fixed.